docs/GDPs/gdp-0003-duck-storage.rst

```diff
@@ -0,0 +1,822 @@
+          :code:`sync_state` attribute. For more details see :ref:`sync_state`.
+
+        For common keyword-only arguments, please see below.
+
+:code:`storage(data: array_like = None, device_data: array_like = None, *, dtype: dtype_like = np.float64, copy=True, **kwargs) -> Storage`
+    Used to allocate a storage with values initialized to those of a given array.
+    If the argument :code:`copy` is set to :code:`False`, the behavior is that of :code:`as_storage`.
+
+    Parameters:
+        + :code:`data: array_like`. The original array from which the storage is initialized.
+
+        + :code:`device_data: array_like`. The original array in case copying to a gpu buffer is
+          desired. The same buffer could also be passed through `data` in that case, however this
+          parameter is here to provide the same interface like the :code:`as_storage` function.
+
+        + :code:`sync_state: gt4py.storage.SyncState`. If `managed="gt4py"` indicates which of the
+          provided buffers, `data` or `device_data`, is up to date at the time of initialization.
+
+    Keyword-only parameters:
+        + :code:`copy: bool`. Allocate a new buffer and initialize it with a copy of the data or
+          wrap the existing buffer.
+
+        + :code:`sync_state: gt4py.storage.SyncState`. If `managed="gt4py"` indicates which of the
+          provided buffers, `data` or `device_data`, is up to date at the time of initialization.
+
+        For common keyword-only arguments, please see below.
+
+    If :code:`copy=False` and neither :code:`data` nor :code:`device_data` are provided, the other
+    arguments are used to allocate an appropriate buffer without initialization (equivalent to call
+    :code:`empty()`). If :code:`data` or :code:`device_data` is provided, the consistency of the
+    parameters with the buffers is validated.
+
+    If the field is not 3-D, as indicated by :code:`axes`, the length of parameters
+    :code:`aligned_index` and :code:`shape`, may either be of length 3 or of the actual dimension
+    of the storage, where the not needed entries are ignored in the latter case.
+
+
+The definitions of the common parameters accepted by all the previous functions are the following:
+
+:code:`dtype: dtype_like`
+    The dtype of the storage (NumPy dtype or accepted by :code:`np.dtype()`). It defaults to
+    :code:`np.float64`.
+
+:code:`shape: Sequence[int]`
+    Sequence of length :code:`ndim` (:code:`ndim` = number of dimensions, maximum 3) with the shape
+    of the storage, that is, the full addressable space in the allocated memory buffer. (See also
+    Section :ref:`domain_and_halo`)
+
+Additionally, these **optional** keyword-only parameters are accepted:
+
+:code:`aligned_index: Sequence[int]`
+    The point to which the memory is aligned, defaults to the lower indices of the halo attribute.
+
+:code:`alignment: Optional[int]`
+    Indicates on a boundary of how many elements the point :code:`aligned_index` is aligned. It
```

Comment:
So, a given point in the grid is aligned to a certain alignment size. (we could change the name, I see how it can be confusing)

So, if the memory address of the grid point denoted by `aligned_index` is `aligned_addr`, then it is considered aligned if `mod(aligned_addr,alignment)==0`. This is a performance feature. If you always go through the `defaults` parameter and specify your halo as the value that is actually used in the computation, you never have to worry about this.


Otherwise, you may or may not suffer a performance penalty.